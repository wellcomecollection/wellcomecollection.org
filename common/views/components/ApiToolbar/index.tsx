import { FunctionComponent, useState } from 'react';
import styled from 'styled-components';
import cookies from '@weco/common/data/cookies';
import { getCookie, setCookie } from 'cookies-next';
import useIsomorphicLayoutEffect from '../../../hooks/useIsomorphicLayoutEffect';
import { Contributor, License } from '../../../model/catalogue';

export type ApiToolbarLink = {
  id: string;
  label: string;
  value?: string;
  link?: string;
};

const ToolbarContainer = styled.div<{ mini: boolean }>`
  display: ${props => (props.mini ? 'inline-block' : 'flex')};
  background-color: ${props => props.theme.color('accent.purple')};
  color: ${props => props.theme.color('white')};
  z-index: 100;
`;

const LinkList = styled.ul.attrs({
  className: 'font-size-5',
})`
  display: flex;
  list-style: none;
  margin: 0;
`;

/** tzitzit is a tool used by the Digital Editorial team to create TASL
 * strings for Prismic. https://github.com/wellcomecollection/tzitzit
 *
 * This creates a link to a tzitzit which is pre-filled with the catalogue
 * metadata for this image.  They may not want to use it directly
 * (e.g. they may tidy up the title), but it should save them some copy/pasting
 * from the catalogue.
 *
 */

export function setTzitzitParams({
  title,
  sourceLink,
  licence,
  contributors,
}: {
  title: string;
  sourceLink: string;
  licence: License | undefined;
  contributors: Contributor[] | undefined;
}): ApiToolbarLink | undefined {
  const licenceId = licence?.id.toUpperCase();

  // We should not be using in copyright images in Stories.
  // TODO: Double check this is our policy with the Editorial team.
  if (licenceId === 'INC')
    return {
      id: 'tzitzit',
      label: '(no tzitzit, this image is in copyright)',
    };

  const params = new URLSearchParams();
  params.set('title', title);
  params.set('sourceName', 'Wellcome Collection');
  params.set('sourceLink', sourceLink);
  if (licenceId) params.set('licence', licenceId);
  if (contributors && contributors.length > 0)
    params.set('author', contributors[0].agent.label);

  return {
    id: 'tzitzit',
    label: 'tzitzit',
    link: `https://s3-eu-west-1.amazonaws.com/tzitzit.wellcomecollection.org/index.html?${params.toString()}`,
  };
}

export function createPrismicLink(prismicId: string): ApiToolbarLink {
  return {
    id: 'prismic',
    label: 'Edit in Prismic',
    link: `https://wellcomecollection.prismic.io/documents~b=working&c=published&l=en-gb/${prismicId}/`,
  };
}

function getAnchorLinkUrls() {
  // This function currently only extracts the ids from h2, h3, and h4 tags
  const getAllHeadingIds = [...document.querySelectorAll('h2, h3, h4')].map(
    item => item.id
  );
  // This function extracts any data-toolbar-anchor divs with the view to extracting the id values
  // This can be used across the codebase (where data-toolbar-anchor is used) but at the moment is only used in audio & BSL guides
  // Please note: an audio/BSL guide must contain and audio or video file with an accompanying title or no id will exist
  // and no link will be created
  const extractedAudioBSLAttributes = [
    ...document.querySelectorAll('div[data-toolbar-anchor="apiToolbar"]'),
  ].map(item => item.id);

  // Remove empty ids and then append them to the current url with # to
  // create the anchor link
  // e.g. weco.org/guides/exhibitions/YvUALRAAACMA2h8V/captions-and-transcripts#anchor-id
  const extractedHeadingIdURLs = getAllHeadingIds
    .filter(Boolean)
    .map(id => `${document.URL}#${id}`);
  const extractedAudioBSLURLs = extractedAudioBSLAttributes
    .filter(Boolean)
    .map(id => `${document.URL}#${id}`);

  const csvAsSingleColumn =
    extractedHeadingIdURLs.join('\n') + extractedAudioBSLURLs.join('\n');
  // Push the list of urls to the clipboard
  if (navigator && navigator.clipboard && navigator.clipboard.writeText)
    return navigator.clipboard.writeText(csvAsSingleColumn).then(() => {
      alert('All anchor links on this page have been copied to clipboard!');
    });
  return alert(
    'The Clipboard API is not available. No anchor links have been copied.'
  );
}

type Props = {
  links?: ApiToolbarLink[];
};

const ApiToolbar: FunctionComponent<Props> = ({ links = [] }) => {
  const [mini, setMini] = useState<boolean>(false);

  useIsomorphicLayoutEffect(() => {
    setMini(getCookie(cookies.apiToolbarMini) === true);
  }, []);

  const propValue = (prop: ApiToolbarLink) => {
    return `${prop.label}${prop.value ? ` : ${prop.value}` : ''}`;
  };

  return (
    <ToolbarContainer mini={mini}>
      <div
        style={{
          display: 'flex',
          alignItems: 'center',
          flexGrow: 1,
        }}
      >
        {!mini && (
          <>
            <span
              className="h3"
              style={{
                marginLeft: '10px',
              }}
            >
              API toolbar
            </span>
            <LinkList>
              {links.map(prop => (
                <li
                  key={prop.id}
                  style={{
                    paddingRight: '10px',
                    paddingLeft: '10px',
                    borderLeft: '1px solid #bcbab5',
                  }}
                >
                  {prop.link && <a href={prop.link}>{propValue(prop)}</a>}
                  {!prop.link && propValue(prop)}
                </li>
              ))}
            </LinkList>
          </>
        )}
      </div>
      <button
        className="plain-button"
        type="button"
        onClick={() => {
          getAnchorLinkUrls();
        }}
        style={{ padding: '10px' }}
      >
        <>‚öìÔ∏è</>
      </button>

      <button
        className="plain-button"
        type="button"
        onClick={() => {
          setMini(!mini);
          setCookie(cookies.apiToolbarMini, !mini, { path: '/' });
        }}
        style={{ padding: '10px' }}
      >
        {!mini && <>ü§è</>}
        {mini && <>üëê</>}
      </button>
    </ToolbarContainer>
  );
};

export default ApiToolbar;
