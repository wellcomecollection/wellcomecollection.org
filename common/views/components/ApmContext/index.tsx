import { ApmBase } from '@elastic/apm-rum';
import getConfig from 'next/config';
import { createContext, FunctionComponent, PropsWithChildren } from 'react';

import useApmRum from './useApmRum';

type ApmContextData = {
  apm?: ApmBase;
};

export const ApmContext = createContext<ApmContextData>({});

const { publicRuntimeConfig } = getConfig();

export const ApmContextProvider: FunctionComponent<PropsWithChildren> = ({
  children,
}) => {
  const apm = useApmRum(publicRuntimeConfig.apmConfig);
  return <ApmContext.Provider value={{ apm }}>{children}</ApmContext.Provider>;
};
