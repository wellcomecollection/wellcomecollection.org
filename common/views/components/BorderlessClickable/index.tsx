import {
  ComponentProps,
  forwardRef,
  ForwardRefRenderFunction,
  ReactNode,
  SyntheticEvent,
} from 'react';
import styled from 'styled-components';

import { IconSvg } from '@weco/common/icons';
import { classNames, font } from '@weco/common/utils/classnames';
import {
  BaseButtonInner,
  BasicButton,
  ButtonIconWrapper,
} from '@weco/common/views/components/Buttons/Buttons.styles';
import Icon from '@weco/common/views/components/Icon';

type ClickableElement = 'a' | 'button';

type StyleProps = {
  $isActive?: boolean;
};

const BorderlessClickableStyle = styled(BasicButton)<StyleProps>`
  background: transparent;
  color: ${props => props.theme.color('neutral.700')};
  padding: 10px 8px;

  &:hover {
    background: ${props => props.theme.color('neutral.300')};
  }

  ${props =>
    props.$isActive &&
    `
    background: ${props.theme.color('neutral.300')};
  `}
`;

type Props = {
  icon?: IconSvg;
  iconLeft?: IconSvg;
  text: ReactNode;
  isTextHidden?: boolean;
  isActive?: boolean;
};

type BorderlessClickableProps = Props & { as: ClickableElement };
const Button: ForwardRefRenderFunction<
  HTMLButtonElement,
  BorderlessClickableProps
> = (
  {
    as,
    icon,
    iconLeft,
    text,
    isTextHidden,
    isActive,
    ...elementProps
  }: BorderlessClickableProps,
  ref
) => {
  return (
    <BorderlessClickableStyle
      as={as}
      $isActive={isActive}
      ref={ref}
      {...elementProps}
    >
      <BaseButtonInner $isInline={true}>
        <>
          {iconLeft && (
            <ButtonIconWrapper $iconAfter={false}>
              {/* This is all a little hacky and will need some tidy up */}
              {/* We currently only use this in the header sign in button */}
              <span
                className={font('intr', 4)}
                style={{ transform: 'translateY(0.01em)' }}
              >
                <Icon icon={iconLeft} matchText={true} />
              </span>
            </ButtonIconWrapper>
          )}
          <span
            className={classNames({
              'visually-hidden': !!isTextHidden,
            })}
          >
            {text}
          </span>
          {icon && (
            <ButtonIconWrapper $iconAfter={true}>
              <Icon icon={icon} />
            </ButtonIconWrapper>
          )}
        </>
      </BaseButtonInner>
    </BorderlessClickableStyle>
  );
};

const BorderlessClickable = forwardRef<
  HTMLButtonElement,
  BorderlessClickableProps
>(Button);

type BorderlessLinkProps = Props & ComponentProps<'a'>;
const Link = (props, ref) => {
  return <BorderlessClickable as="a" ref={ref} {...props} />;
};

const BorderlessLink = forwardRef<HTMLButtonElement, BorderlessLinkProps>(Link);

type BorderlessButtonProps = Props &
  ComponentProps<'button'> & {
    clickHandler?: (event: SyntheticEvent<HTMLButtonElement>) => void;
  };
const ButtonOuter = (
  { clickHandler, ...elementProps }: BorderlessButtonProps,
  ref
) => {
  function onClick(event: SyntheticEvent<HTMLButtonElement>) {
    clickHandler && clickHandler(event);
  }

  return (
    <BorderlessClickable
      as="button"
      onClick={onClick}
      ref={ref}
      {...elementProps}
    />
  );
};

const BorderlessButton = forwardRef<HTMLButtonElement, BorderlessButtonProps>(
  ButtonOuter
);

export { BorderlessLink, BorderlessButton };
