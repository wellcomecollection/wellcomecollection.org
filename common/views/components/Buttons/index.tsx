import {
  ForwardRefRenderFunction,
  ForwardedRef,
  PropsWithChildren,
  forwardRef,
} from 'react';
import ButtonSolid, { ButtonSolidProps } from './Buttons.Solid';
import ButtonSolidLink, { ButtonSolidLinkProps } from './Buttons.SolidLink';
import DropdownButton, { DropdownButtonProps } from './Buttons.Dropdown';
// import {
//   BorderlessLink,
//   BorderlessButton,
//   BorderlessLinkProps,
//   BorderlessButtonProps,
// } from './Buttons.Borderless';

type ButtonProps =
  // | (BorderlessButtonProps & {
  //     variant: 'BorderlessButton';
  //   })
  // | (BorderlessLinkProps & {
  //     variant: 'BorderlessLink';
  //   })
  | (PropsWithChildren<DropdownButtonProps> & {
      variant: 'DropdownButton';
    })
  | (ButtonSolidProps & {
      variant: 'ButtonSolid';
    })
  | (ButtonSolidLinkProps & {
      variant: 'ButtonSolidLink';
    });

const Button: ForwardRefRenderFunction<HTMLButtonElement, ButtonProps> = (
  props: ButtonProps,
  ref?: ForwardedRef<HTMLButtonElement>
) => {
  const { variant } = props;

  switch (variant) {
    // case 'BorderlessButton':
    //   return <BorderlessButton {...props} />;
    // case 'BorderlessLink':
    //   return <BorderlessLink {...props} />;
    case 'DropdownButton':
      return <DropdownButton {...props} />;
    case 'ButtonSolidLink':
      return <ButtonSolidLink {...props} />;
    case 'ButtonSolid':
    default:
      return <ButtonSolid ref={ref} {...props} />;
  }
};

export default forwardRef<HTMLButtonElement, ButtonProps>(Button);
export * from './Buttons.styles';
export * from './Buttons.types';
