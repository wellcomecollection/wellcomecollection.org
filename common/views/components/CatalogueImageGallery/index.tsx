import { FunctionComponent } from 'react';

import CatalogueImageGalleryJustified, {
  Props as CatalogueImageGalleryJustifiedProps,
} from '@weco/common/views/components/CatalogueImageGallery/CatalogueImageGallery.Justified';
import CatalogueImageGalleryScrollable, {
  Props as CatalogueImageGalleryScrollableProps,
} from '@weco/common/views/components/CatalogueImageGallery/CatalogueImageGallery.Scrollable';

type Props =
  | (CatalogueImageGalleryJustifiedProps & { variant: 'justified' })
  | (CatalogueImageGalleryScrollableProps & { variant: 'scrollable' });

const CatalogueImageGallery: FunctionComponent<Props> = props => {
  const { variant } = props;

  return (
    <>
      {variant === 'justified' ? (
        <CatalogueImageGalleryJustified {...props} />
      ) : (
        <CatalogueImageGalleryScrollable {...props} />
      )}
    </>
  );
};

export default CatalogueImageGallery;
