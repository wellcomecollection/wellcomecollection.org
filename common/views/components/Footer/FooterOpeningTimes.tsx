import { ReactElement } from 'react';
import { font } from '@weco/common/utils/classnames';
import Space from '@weco/common/views/components/styled/Space';
import OpeningTimes from '@weco/common/views/components/OpeningTimes/OpeningTimes';
import { Venue } from '@weco/common/model/opening-hours';

const FooterOpeningTimes = ({ venues }: { venues: Venue[] }): ReactElement => {
  return (
    <>
      <h4 className={font('intb', 5)}>{`Today's opening times`}</h4>
      {venues && <OpeningTimes venues={venues} />}
      <Space as="p" v={{ size: 's', properties: ['margin-top'] }}>
        <a href="/opening-times">Opening times</a>
      </Space>
    </>
  );
};

export default FooterOpeningTimes;
