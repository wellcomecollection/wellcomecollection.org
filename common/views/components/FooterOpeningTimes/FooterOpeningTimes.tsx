import { getTodaysVenueHours } from '@weco/common/services/prismic/opening-times';
import Space from '../styled/Space';
import { CollectionOpeningTimes } from '@weco/common/model/opening-hours';
import {
  collectionVenueId,
  getNameFromCollectionVenue,
} from '@weco/common/services/prismic/hardcoded-id';
import { FunctionComponent, ReactElement } from 'react';

type Props = {
  collectionOpeningTimes: CollectionOpeningTimes;
};

const FooterOpeningTimes: FunctionComponent<Props> = ({
  collectionOpeningTimes,
}: Props): ReactElement<Props> => {
  return (
    <ul className="plain-list no-padding no-margin">
      {collectionOpeningTimes.placesOpeningHours.map(venue => {
        const todaysHours = getTodaysVenueHours(venue);
        return (
          todaysHours && (
            <Space
              v={{
                size: 's',
                properties: ['margin-top'],
              }}
              as="li"
              key={venue.id}
            >
              {venue.id === collectionVenueId.restaurant.id
                ? 'Kitchen '
                : `${getNameFromCollectionVenue(venue.id)} `}
              {todaysHours.opens ? (
                <>
                  <time>{todaysHours.opens}</time>
                  {'â€”'}
                  <time>{todaysHours.closes}</time>
                </>
              ) : (
                'closed'
              )}
            </Space>
          )
        );
      })}
    </ul>
  );
};
export default FooterOpeningTimes;
