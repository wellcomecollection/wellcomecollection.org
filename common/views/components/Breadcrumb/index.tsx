import { FunctionComponent } from 'react';

import { BreadcrumbItems } from '@weco/common/model/breadcrumbs';
import { classNames, font } from '@weco/common/utils/classnames';
import { breadcrumbsLd } from '@weco/common/utils/json-ld';
import { links } from '@weco/common/views/components/Header';
import Space from '@weco/common/views/components/styled/Space';

export function getBreadcrumbItems(siteSection?: string): BreadcrumbItems {
  if (!siteSection) return { items: [] };
  return {
    items: siteSection
      ? [
          {
            text:
              links.find(link => link.siteSection === siteSection)?.title ||
              siteSection,
            url: siteSection ? `/${siteSection}` : '',
          },
        ]
      : [],
  };
}

const Breadcrumb: FunctionComponent<BreadcrumbItems> = ({
  items,
  noHomeLink,
}) => {
  // We prepend a 'Home' breadcrumb at the start of every chain, so every page
  // will ideally always have a visible breadcrumb.
  const allItems = [
    {
      text: 'Home',
      url: '/',
      isHidden: noHomeLink,
    },
    ...items,
  ];

  const visibleItems = allItems.filter(({ isHidden }) => !isHidden);

  if (visibleItems.length === 0) return null;

  return (
    // TODO remove is-hidden-print class once we've made the breadcrumbs more useful
    <div className="is-hidden-print" data-testid="breadcrumbs">
      {visibleItems.map(({ text, url, prefix }, i) => {
        const LinkOrSpanTag = url ? 'a' : 'span';

        return (
          <Space
            key={prefix ? `${prefix}-${text}` : text}
            as={prefix ? 'b' : 'span'}
            className={font('intr', 6)}
          >
            {i > 0 && (
              <Space
                as="span"
                $h={{ size: 's', properties: ['margin-left', 'margin-right'] }}
                aria-hidden="true"
              >
                |
              </Space>
            )}
            {prefix}{' '}
            <LinkOrSpanTag
              className={classNames({
                [font('intb', 6)]: Boolean(prefix),
              })}
              href={url}
              data-gtm-trigger={url ? 'breadcrumb_link' : undefined}
            >
              {text}
            </LinkOrSpanTag>
          </Space>
        );
      })}
      <script
        type="application/ld+json"
        dangerouslySetInnerHTML={{
          __html: JSON.stringify(breadcrumbsLd({ items })),
        }}
      />
    </div>
  );
};
export default Breadcrumb;
