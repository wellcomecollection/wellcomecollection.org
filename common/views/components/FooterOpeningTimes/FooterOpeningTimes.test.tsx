import FooterOpeningTimes from './FooterOpeningTimes';
import {
  shallowWithTheme,
  mountWithTheme,
} from '../../../test/fixtures/enzyme-helpers';
import { openingTimes } from '../../../test/fixtures/components/opening-times';
import * as serviceOpeningTimes from '@weco/common/services/prismic/opening-times';

describe('FooterOpeningTimes', () => {
  const spyOnGetTodaysVenueHours = jest.spyOn(
    serviceOpeningTimes,
    'getTodaysVenueHours'
  );
  // set date as wednesday
  spyOnGetTodaysVenueHours.mockImplementation(() => {
    return { dayOfWeek: 'Wednesday', opens: '10:00', closes: '18:00' };
  });

  it('Should display opening times name restaurant as Kitchen', () => {
    const component = shallowWithTheme(
      <FooterOpeningTimes
        collectionOpeningTimes={openingTimes.collectionOpeningTimes}
      />
    );

    expect(component.html().match('Kitchen '));
  });

  it('Should not display render any opening times if opening times are empty', () => {
    const mockOpeningTimes = {
      collectionOpeningTimes: {
        placesOpeningHours: [
          {
            id: 'WsuaIB8AAH-yNylo',
            order: 5,
            name: 'Shop',
            openingHours: {
              regular: [],
              exceptional: [],
            },
          },
        ],
      },
    };
    spyOnGetTodaysVenueHours.mockImplementation(() => {
      return null;
    });

    const component = mountWithTheme(
      <FooterOpeningTimes
        collectionOpeningTimes={mockOpeningTimes.collectionOpeningTimes}
      />
    );

    const footerOpeningTimes = component.find('ul');
    expect(footerOpeningTimes.length).toEqual(1);
    expect(footerOpeningTimes.find('li').length).toEqual(0);
  });

  it('Should render venue opening times as closed', () => {
    const mockOpeningTimes = {
      collectionOpeningTimes: {
        placesOpeningHours: [
          {
            id: 'WsuaIB8AAH-yNylo',
            order: 5,
            name: 'Shop',
            openingHours: {
              regular: [
                { dayOfWeek: 'Monday', opens: null, closes: null },
                { dayOfWeek: 'Tuesday', opens: null, closes: null },
                { dayOfWeek: 'Wednesday', opens: null, closes: null },
                { dayOfWeek: 'Thursday', opens: null, closes: null },
                { dayOfWeek: 'Friday', opens: null, closes: null },
                { dayOfWeek: 'Saturday', opens: null, closes: null },
                { dayOfWeek: 'Sunday', opens: null, closes: null },
              ],
              exceptional: [],
            },
          },
        ],
      },
    };

    spyOnGetTodaysVenueHours.mockImplementation(() => {
      return { dayOfWeek: 'Wednesday', opens: null, closes: null };
    });
    const component = shallowWithTheme(
      <FooterOpeningTimes
        collectionOpeningTimes={mockOpeningTimes.collectionOpeningTimes}
      />
    );
    expect(component.html().match('Shop closed')).toBeTruthy();
  });
});
