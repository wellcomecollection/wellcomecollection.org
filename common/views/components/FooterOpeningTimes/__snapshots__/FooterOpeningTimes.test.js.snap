// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FooterOpeningTimes Should match snapshot of FooterOpeningTimes 1`] = `"<ul class=\\"plain-list no-padding no-margin\\"><li class=\\"Space-sc-4az8w3-0 gmFyho\\">Galleries <time>10:00</time>—<time>18:00</time></li><li class=\\"Space-sc-4az8w3-0 gmFyho\\">Library <time>10:00</time>—<time>18:00</time></li><li class=\\"Space-sc-4az8w3-0 gmFyho\\">Kitchen <time>10:00</time>—<time>18:00</time></li><li class=\\"Space-sc-4az8w3-0 gmFyho\\">Café <time>10:00</time>—<time>18:00</time></li><li class=\\"Space-sc-4az8w3-0 gmFyho\\">Shop <time>10:00</time>—<time>18:00</time></li></ul>"`;
